### ISSUE TITLE
I can assign a string to an Int32 record field in a nested path lvalue

### ISSUE THREAD

**Issue by ckp95** [austral/austral#381]
This compiles when it clearly shouldn't:

```
module body Mutate is
      record Other: Linear is
          y: Nat32;
      end;

      record Mut: Linear is
          x: Other;
      end;

      generic [R: Region]
      function mangle(thing: &![Mut, R]): Unit is
          thing->x.y := "this shouldn't work";
          return nil;
      end;

      function main(): ExitCode is
          return ExitSuccess();
      end;
end module body.
```

The function never gets called but `thing->x.y := "this shouldn't work";` should never be allowed because `.y` is `Nat32`.

If I call it in `main`, Austral doesn't mind. It just makes me destructure `foo`, whereupon `gcc` crashes:

```
module body Mutate is
      record Other: Linear is
          y: Nat32;
      end;

      record Mut: Linear is
          x: Other;
      end;

      generic [R: Region]
      function mangle(thing: &![Mut, R]): Unit is
          thing->x.y := "this shouldn't work";
          return nil;
      end;

      function main(): ExitCode is
          let foo: Mut := Mut(x => Other(y => 5));
          mangle(&!foo);
          let { x: Other } := foo;
          let { y: Nat32 } := x;
          return ExitSuccess();
      end;
end module body.
```

```
$ ./austral compile mine/mutate.aum --entrypoint=Mutate:main --output=mine/mutate
Compiler call tree printed to calltree.html
Error:
  Title: Generic Error
  Module:
    [unknown]
  Location:
    [no span available]
  Description:
    Error when running C compiler.
      Command: cc /tmp/austral_a23ce5.c -fwrapv -lm -o mine/mutate
      Exit code: 1
      Standard output:
    
      Standard error:
    /tmp/austral_a23ce5.c: In function 'mono_15':
    /tmp/austral_a23ce5.c:3969:22: error: incompatible types when assigning to type 'au_nat32_t' {aka 'unsigned int'} from type 'au_array_t'
     3969 |     ((thing)->x).y = au_make_array_from_string("this shouldn\'t work", 19);
          |                      ^~~~~~~~~~~~~~~~~~~~~~~~~
    
  Code:
    [no span available]
```

I found this by accident while experimenting with the path syntax. I was trying to find a way to have the `mangle` function update the value of `y`, on one line.

```
thing->x->y := 10;
```

^ this gave the error `Not a record type`, which I don't understand. After a lot of trial and error I found:

```
thing->x.y := 10;
```

works. It updates the value of `.y`, the code compiles, and I used `printLn` to verify it. But I couldn't figure out from reading the spec when I'm supposed to use `->` and when I'm supposed to use `.` in a path. Especially because this previous attempt *also* works:

```
let z: &![Other, R] := thing->x;
z->y := 10;
```

I don't understand why this one works with `->y` rather than `.y`.

The spec says:

```
    A slot accessor of the form .s where s is the name of a record slot.
    A pointer slot accessor of the form ->s where s is the name of a record slot.
    An array indexing element of the form [i] where i is a value of type Index.
```

But it doesn't explain the distinction between a "slot" and a "pointer slot" .. I don't see anything pointer-y in the above code. I guess it's named in analogy to C's syntax. Does it just mean linear vs free?

**Comment by eudoxia0** [comment]
The error messages for path type errors are not great.

>But it doesn't explain the distinction between a "slot" and a "pointer slot" .. I don't see anything pointer-y in the above code. I guess it's named in analogy to C's syntax. Does it just mean linear vs free?

The arrow operator is for going down references. Pointer dereferencing is done through its own API.

**Comment by eudoxia0** [comment]
Added a test here: #386

**Comment by eudoxia0** [comment]
Fixed by #386. Turns out I forgot type checking in assignments :nauseated_face:
